pkg/integrations/v4/dm/id_provider.go

```diff
@@ -0,0 +1,102 @@
+package dm
+
+import (
+	"fmt"
+	"github.com/newrelic/infrastructure-agent/pkg/backend/identityapi"
+	"github.com/newrelic/infrastructure-agent/pkg/entity"
+	"github.com/newrelic/infrastructure-agent/pkg/integrations/v4/protocol"
+)
+
+type RegisteredEntitiesNameToID map[string]entity.ID
+type UnregisteredEntitiesNamed map[string]UnregisteredEntity
+type reason string
+
+type UnregisteredEntity struct {
+	Reason reason
+	Err    error
+	Entity protocol.Entity
+}
+
+type UnregisteredEntities []UnregisteredEntity
+
+const reasonClientError = "Identity client error"
+const reasonEntityError = "Entity error"
+
+func newUnregisteredEntity(entity protocol.Entity, reason reason, err error) UnregisteredEntity {
+	return UnregisteredEntity{
+		Entity: entity,
+		Reason: reason,
+		Err:    err,
+	}
+}
+
+type idProviderInterface interface {
+	Entities(agentIdn entity.Identity, entities []protocol.Entity) (registeredEntities RegisteredEntitiesNameToID, unregisteredEntities UnregisteredEntities)
+}
+
+type idProvider struct {
+	client               identityapi.RegisterClient
+	cache                RegisteredEntitiesNameToID
+	unregisteredEntities UnregisteredEntitiesNamed
+}
+
+func NewIDProvider(client identityapi.RegisterClient) *idProvider {
+	cache := make(RegisteredEntitiesNameToID)
+	unregisteredEntities := make(UnregisteredEntitiesNamed)
+	return &idProvider{
+		client:               client,
+		cache:                cache,
+		unregisteredEntities: unregisteredEntities,
+	}
+}
+
+func (p *idProvider) Entities(agentIdn entity.Identity, entities []protocol.Entity) (registeredEntities RegisteredEntitiesNameToID, unregisteredEntities UnregisteredEntities) {
+	unregisteredEntities = make(UnregisteredEntities, 0)
+	registeredEntities = make(RegisteredEntitiesNameToID, 0)
+	entitiesToRegister := make([]protocol.Entity, 0)
+
+	for _, e := range entities {
+		if foundID, ok := p.cache[e.Name]; ok {
+			registeredEntities[e.Name] = foundID
+		} else {
+			entitiesToRegister = append(entitiesToRegister, e)
+		}
+	}
+
+	if len(entitiesToRegister) == 0 {
+		return registeredEntities, unregisteredEntities
+	}
+
+	response, _, errClient := p.client.RegisterBatchEntities(
+		agentIdn.ID,
+		entitiesToRegister)
+
+	type nameToEntityType map[string]protocol.Entity
+	nameToEntity := make(nameToEntityType, len(entitiesToRegister))
+
+	for i := range entitiesToRegister {
+		nameToEntity[entitiesToRegister[i].Name] = entitiesToRegister[i]
+	}
+
+	if errClient != nil {
+		for i := range entitiesToRegister {
+			unregisteredEntity := newUnregisteredEntity(entitiesToRegister[i], reasonClientError, errClient)
+			p.unregisteredEntities[entitiesToRegister[i].Name] = unregisteredEntity
+			unregisteredEntities = append(unregisteredEntities, unregisteredEntity)
+		}
+	} else {
+		for i := range response {
+			if response[i].Err != "" {
+				unregisteredEntity := newUnregisteredEntity(nameToEntity[response[i].Name], reasonEntityError, fmt.Errorf(response[i].Err))
```

Comment:
what's the point for generating this error object from string?
isn't its only purpose to be displayed?

Reply:
This is a bit of work in advance, once we will have async part, we will have different reasons to not register entity. 
- `not in cache, sent to the queue to register, come back later`
- `not in cache, http error in client, will try to reregister, come back later`
- `not in cache, http respond with 400, don't ask me again about this entity`
- maybe something else 

Also we can use this for retryBackOff logic.